In [54]:
url = "http://datasystems.denison.edu/basic.html"

In [55]:
import requests
response = requests.get(url)
request = response.request

In [56]:
print("Response status:", response.status_code)

Response status: 200


In [57]:
print("URL:", request.url)

URL: http://datasystems.denison.edu/basic.html


In [58]:
print(response.text, type(response.text), response.content, type(response.content))

<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Data Systems Basic HTML Page</title>
  </head>
  <body>
    <h1>First Level Heading</h1>

    <p>Paragraph defined in <b>body</b>.

    <h2>Second Level Heading</h2>

    <a href="http://docs.python.org">Link</a> to Python documentation.
    </p>

    <ul>
      <li>Item 1
      <ol>
        <li>Item 1 nested</li>
        <li>Item 2 nested</li>
      </ol>
      </li>
      <li>Item 2</li>
      <li>Item 3</li>
    </ul>
  </body>
</html>
 <class 'str'> b'<!DOCTYPE html>\n<html lang="en">\n  <head>\n    <title>Data Systems Basic HTML Page</title>\n  </head>\n  <body>\n    <h1>First Level Heading</h1>\n\n    <p>Paragraph defined in <b>body</b>.\n\n    <h2>Second Level Heading</h2>\n\n    <a href="http://docs.python.org">Link</a> to Python documentation.\n    </p>\n\n    <ul>\n      <li>Item 1\n      <ol>\n        <li>Item 1 nested</li>\n        <li>Item 2 nested</li>\n      </ol>\n      </li>\n      <li>Item 2</li>\n      <li>Item 3</

In [23]:
url = "http://httpbin.org/get"
headerD = {
"Accept": "application/json",
"User-Agent": "datasystems-client"
}
response = requests.get(url, headers=headerD)
request = response.request
print("Response status:", response.status_code)


Response status: 200


In [24]:
response.request.headers

{'User-Agent': 'datasystems-client', 'Accept-Encoding': 'gzip, deflate', 'Accept': 'application/json', 'Connection': 'keep-alive'}

In [25]:
response.request.path_url

'/get'

In [26]:
response.request.url

'http://httpbin.org/get'

In [27]:
response.request.method

'GET'

In [28]:
response.content

b'{\n  "args": {}, \n  "headers": {\n    "Accept": "application/json", \n    "Accept-Encoding": "gzip, deflate", \n    "Host": "httpbin.org", \n    "User-Agent": "datasystems-client", \n    "X-Amzn-Trace-Id": "Root=1-617be5ad-428944ae4ffb7fce2af0f1e5"\n  }, \n  "origin": "14.224.28.87", \n  "url": "http://httpbin.org/get"\n}\n'

In [31]:
print(type(response.content))

<class 'bytes'>


get with query param


In [15]:
url = "http://httpbin.org/get"
paramsD = {
"user": "smith",
"query": "movies tv"
}
response = requests.get(url, params=paramsD)
request = response.request
print("Response status:", response.status_code)

Response status: 200


In [16]:
print("Path:", request.path_url)

Path: /get?user=smith&query=movies+tv


post

In [47]:
url = "http://httpbin.org/post"
paramsD = {
"user": "jones",
"query": "TV?episodes"
}
headerD = {
"Accept": "application/json"
}
body = {"a": 1, "b": 2}
response = requests.post(url, params=paramsD,
headers=headerD, data=body)
request = response.request
print("Response status:", response.status_code)

Response status: 200


In [48]:
print("Request Path:", request.path_url)

Request Path: /post?user=jones&query=TV%3Fepisodes


In [49]:
print("Request Body:", request.body)

Request Body: a=1&b=2


In [51]:
print(type(response.content))

<class 'bytes'>


post with json body

In [36]:
import json
url = "http://httpbin.org/post"
paramsD = {
"user": "jones",
"query": "TV"
}
headerD = {
"Accept": "application/json"
}
json_data = ["foo", "bar", {"a": 1, "b": 2}]
response = requests.post(url, params=paramsD,
headers=headerD, json=json_data)
request = response.request
print("Response status:", response.status_code)

Response status: 200


In [37]:
print("Request Path:", request.path_url)

Request Path: /post?user=jones&query=TV


In [38]:
print("Request Body:", request.body.decode('utf-8'))

Request Body: ["foo", "bar", {"a": 1, "b": 2}]


In [52]:
response.text

'{\n  "args": {\n    "query": "TV?episodes", \n    "user": "jones"\n  }, \n  "data": "", \n  "files": {}, \n  "form": {\n    "a": "1", \n    "b": "2"\n  }, \n  "headers": {\n    "Accept": "application/json", \n    "Accept-Encoding": "gzip, deflate", \n    "Content-Length": "7", \n    "Content-Type": "application/x-www-form-urlencoded", \n    "Host": "httpbin.org", \n    "User-Agent": "python-requests/2.26.0", \n    "X-Amzn-Trace-Id": "Root=1-617be651-49f600bc418620a3304637e5"\n  }, \n  "json": null, \n  "origin": "14.224.28.87", \n  "url": "http://httpbin.org/post?user=jones&query=TV%3Fepisodes"\n}\n'

In [53]:
response.content

b'{\n  "args": {\n    "query": "TV?episodes", \n    "user": "jones"\n  }, \n  "data": "", \n  "files": {}, \n  "form": {\n    "a": "1", \n    "b": "2"\n  }, \n  "headers": {\n    "Accept": "application/json", \n    "Accept-Encoding": "gzip, deflate", \n    "Content-Length": "7", \n    "Content-Type": "application/x-www-form-urlencoded", \n    "Host": "httpbin.org", \n    "User-Agent": "python-requests/2.26.0", \n    "X-Amzn-Trace-Id": "Root=1-617be651-49f600bc418620a3304637e5"\n  }, \n  "json": null, \n  "origin": "14.224.28.87", \n  "url": "http://httpbin.org/post?user=jones&query=TV%3Fepisodes"\n}\n'

In [45]:
print(type(response.text))

<class 'str'>


**Exercises**

**20.35** The requests module uses URLs as the first argument to its HTTP method
functions, but we often start with the “piece parts” of the information contained in
a URL. Write a function
buildURL(resource, location, protocol=’http’)
that returns a string URL based on the three component parts of protocol,
location, and resource. Your function should be flexible, so that if a user
omits a leading \ on the resource path, one is prepended. Note that we are specifying
a default value for protocol so that it will use http if buildURL is called with
just two arguments. The string method format() is the right tool for the job here.

In [8]:
def buildURL(resource, location, protocol='http'):
    return '{}://{}{}'.format(protocol, location, resource)


location = 'personal.denison.edu'
resource = '/~kretchmar/'
print(buildURL(resource, location))

http://personal.denison.edu/~kretchmar/


**20.36** Write a sequence of code that starts with:
resource = "/data/ind0.json"
location = "datasystems.denison.edu"
and build an appropriate URL, uses requests to issue a GET request, and
assigns the variables based on the result:
- status: has the integer status code,
- headers: has a dictionary of headers from the response, and
- data has the parsed data from the JSON-formatted body

In [11]:
resource = "/data/ind0.json" 
location = "datasystems.denison.edu" 
url = buildURL(resource, location)

response = requests.get(url)
request = response.request


In [14]:
print(url)
print(response.status_code)
print(response.headers)
print(response.content)

http://datasystems.denison.edu/data/ind0.json
200
{'Date': 'Fri, 29 Oct 2021 03:24:07 GMT', 'Server': 'Apache/2.4.6 (CentOS)', 'Last-Modified': 'Wed, 16 Dec 2020 23:45:42 GMT', 'ETag': '"10d-5b69d7922d580"', 'Accept-Ranges': 'bytes', 'Content-Length': '269', 'Connection': 'close', 'Content-Type': 'application/json'}
b'{"FRA": {"2007": {"pop": 64.02, "gdp": 2657.21}, "2017": {"pop": 66.87, "gdp": 2586.29}}, "GBR": {"2007": {"pop": 61.32, "gdp": 3084.12}, "2017": {"pop": 66.06, "gdp": 2637.87}}, "USA": {"2007": {"pop": 301.23, "gdp": 14451.9}, "2017": {"pop": 325.15, "gdp": 19485.4}}}'


In [15]:
data = response.json()
data

{'FRA': {'2007': {'pop': 64.02, 'gdp': 2657.21},
  '2017': {'pop': 66.87, 'gdp': 2586.29}},
 'GBR': {'2007': {'pop': 61.32, 'gdp': 3084.12},
  '2017': {'pop': 66.06, 'gdp': 2637.87}},
 'USA': {'2007': {'pop': 301.23, 'gdp': 14451.9},
  '2017': {'pop': 325.15, 'gdp': 19485.4}}}

**20.37** Suppose you often coded a similar set of steps to make a GET request, where
often the body of the result was JSON, in which case you wanted the data parsed,
but sometimes the data was not JSON, in which case you wanted the data as a string.
Write a function
- makeRequest(resource, location, protocol="http")

that makes a GET request to the given location, resource, and
protocol. If the request is not successful (i.e., not in the 200’s), the function
should check for this and return None. If the request is successful, the function
should use the response headers and determine whether or not the Content-Type
header maps to application/json. If it is, it should parse the result and return
the data structure. If it is not, it should return the string making up the body of the
response.

In [19]:
def makeRequest(resouce, location, protocol="http"):
    def buildURL(resource, location, protocol='http'):
        return '{}://{}{}'.format(protocol, location, resource)
    url = buildURL(resource, location)
    response = requests.get(url)
    status_code = response.status_code
    if status_code == 200:
        if response.headers['Content-Type']== 'application/json':
            return response.json()
        else:
            response.content
    else:
        return None

resource = "/data/ind0.json" 
location = "datasystems.denison.edu" 
data = makeRequest(resource, location)
data

{'FRA': {'2007': {'pop': 64.02, 'gdp': 2657.21},
  '2017': {'pop': 66.87, 'gdp': 2586.29}},
 'GBR': {'2007': {'pop': 61.32, 'gdp': 3084.12},
  '2017': {'pop': 66.06, 'gdp': 2637.87}},
 'USA': {'2007': {'pop': 301.23, 'gdp': 14451.9},
  '2017': {'pop': 325.15, 'gdp': 19485.4}}}

**20.38** You have probably had the experience before of trying to open a web page,
and having a redirect page pop up, telling you that the page has moved and asking if
you want to be redirected. The same thing can happen when we write code to make
requests. Write a function:

- getRedirectURL(resource, location)

that begins like your function makeRequest but does not allow redirects when
invoking get. This function will return a url. 
- If the get results in a success status
code (one in the 200’s), you return the original url (obtained from buildURL, with
http protocol). 

- If you detect that get tried to redirect (by looking for a 300, 301,
or 302 status code), search within the headers to find the "Location" it tried to
redirect to, and return that URL instead. 
- If you get any other status code, return
None.